### ISSUE TITLE
ASR模型微调-优化

### ISSUE THREAD

**Issue by alexanderAngle** [shibing624/pycorrector#593]
### Describe the Question
Please provide a clear and concise description of what the question is.
现在在使用FUNASR里的开源SenseVoiceSmall模型做重庆话方言ASR训练，目前训练的一个结果就是训练确实有效果但是效果不是很好；目前累计的数据大概有50个小时左右的数据，字错率大概在19%左右（特殊的语气词完全没有做SPECIAL_TOKEN处理）；目前在尝试将这些特殊的语气词做SPECIAL_TOKEN处理之后再次训练查看训练的一个效果。除此之外还有其他的方式可以去对训练做一下优化嘛。目前累计的方言数据中部分数据的音频其实并不是很好通常会有一些听不清、杂音等数据；这部分数据较少；除此之外还有其他能够对此做ASR训练调优的方式嘛，我想了解一下其他的思路。（目前只是针对官方提供的脚本对数据集做训练）

**Comment by shibing624** [comment]
# 重庆话 ASR（SenseVoiceSmall）微调优化方案


---

## 0. 先做错误分析，再谈优化（否则全是盲调）

19% 这个数字本身信息量不够。在动任何超参之前，先把错误拆开：

### 0.1 CER 拆解为 Sub / Del / Ins

```python
import jiwer
out = jiwer.process_words(ref, hyp)
# 或用 char-level：jiwer.cer(ref, hyp, return_dict=True)
```

三类错误指向完全不同的根因：

| 主导错误类型 | 典型根因 | 对应手段 |
|---|---|---|
| **Sub（替换）为主** | 同音/近音字混淆、同音异写、方言词无标准写法 | 文本规范化、混淆词表、LM/后纠错 |
| **Del（删除）为主** | 音频质量差、语速快、音量小、VAD 切分截断 | 数据清洗、降噪、VAD 参数、音频筛选 |
| **Ins（插入）为主** | 语气词/重复/口吃、标注未规范、模型复读 | 语气词处理、去重、非自回归模型的重复抑制 |

### 0.2 高频混淆对 Top-N

统计 `ref→hyp` 的对齐结果，取频次 Top 50 的替换对。经验上 **Top 50 混淆对往往能解释 30%+ 的总错误**，且大部分集中在：
- 同音字（"在/再"、"的/地/得"、"做/作"）
- 方言词无标准写法（"啷个/朗个/哪个"、"啥子/撒子/啥子"）
- 方言特有词被写成普通话近音字

### 0.3 分 slice 看 CER

按 **说话人 / 录音设备 / 音频时长 / SNR / 语速（字/秒）** 分组统计 CER。最差的那 20% slice 通常贡献了 50%+ 的错误，且往往是某一类系统性问题（比如某一个录音设备、某一位标注员）。

### 0.4 过拟合判断

对比 **train CER vs val CER**：
- train << val（如 5% vs 19%）→ 严重过拟合，优先降 lr / 加正则 / 减 epoch / 加数据增强
- train ≈ val ≈ 19% → 欠拟合或数据噪声天花板，优先清洗数据、换容量更大的方案

**这一条决定了后面所有优化的方向，必须先做。**

---

## 1. 数据侧（收益最大，也是 50h 场景唯一能拿大头的地方）

### 1.1 标注质量清洗 —— 优先级最高

方言数据集的标注噪声通常远超想象。用当前最好的模型（微调后的模型，或更大的模型如 `paraformer-zh` / `Whisper-large-v3`）对**训练集本身**做一遍推理，算每条的 CER：

```bash
# 对训练集全量推理，得到每条 CER
# CER 分布分层处理：
#   CER > 60%  → 大概率标注错误或音频不可用，直接丢
#   20% ~ 60% → 人工抽检定性：是标注错还是音频差？
#   < 20%      → 保留
```

**关键**：丢之前先人工抽听最差的 5%（约几十条）定性。如果这几十条里大部分是"标注文本本身写错了/错位的"，说明你的 19% 里有一块是**数据集天花板**，再怎么调参都压不下去，必须清洗。

小数据量下，**500 条高质量样本的收益 > 5000 条含噪样本**。

### 1.2 音频质量筛选（针对你提到的"部分音频听不清、有杂音"）

不用人工听全部，用低成本自动指标批量打分：

| 指标 | 做法 | 用途 |
|---|---|---|
| **DNSMOS**（微软开源无参考音质打分） | `pip install onnxruntime` + DNSMOS P.808/P.835 模型 | 直接给出 MOS 分，< 2.5 的单独处理 |
| **SNR 估计** | WebRTC VAD 分帧 → 语音段能量/非语音段能量 | 剔除低 SNR 样本 |
| **字速异常** | `len(text) / duration`，正常中文约 4~7 字/秒 | 过高=漏标或语速异常，过低=多标或静音过长 |
| **有效语音占比** | VAD 语音时长 / 总时长 | 剔除大量静音/噪声段 |

**处理策略（按优先级）**：
1. **量少就别删，先降噪**：用 FunASR 生态自带的降噪模型批量预处理，统一输入分布
   ```python
   from funasr import AutoModel
   ans = AutoModel(model="damo/speech_frcrn_ans_cirm_16k")
   ans.generate(input="in.wav", output_path="out_dir")
   ```
   注意：**训练集和测试集要用同一套预处理**，否则引入训练/推理不一致。
2. **样本加权**：给低质量样本降 loss 权重（如 0.5），而不是直接丢弃，保留说话人多样性
3. **课程学习（Curriculum Learning）**：前期只用干净数据，后期逐步混入脏数据
4. 直接丢弃只用于占比极小且确认不可用的部分

### 1.3 文本规范化（TN）—— 这一步优先级高于 SPECIAL_TOKEN

SenseVoice 输出的是**富文本**（带标点 + ITN），所以训练文本必须和推理期望的输出格式严格一致。常见坑：

- **数字/日期/金额/英文**：统一书写形式（"2024年" vs "二零二四年"、"3块" vs "三块"）。SenseVoice 有 `with_or_wo_itn` 标签，训练时如果用 `<|withitn|>`，文本就必须是 ITN 后的形式。
- **标点一致性**：要么全加标点，要么全不加，**不能混**。如果标注源没有标点，建议统一去掉，并把 `with_or_wo_itn` 设为 `<|woitn|>`。
- **繁简/异体字**：统一到简体规范字。

**这一条如果没有做，会直接吃掉几个点的 CER**，因为模型一半的容量在学"到底该输出哪种写法"。

### 1.4 同音异写收敛 —— 对重庆话尤其重要

重庆话大量词汇没有标准书面形式，标注员各自为政：

```
啥子 / 撒子 / 啥子  →  统一为「啥子」
啷个 / 朗个 / 浪个  →  统一为「啷个」
莫得 / 没得 / 木得  →  统一为「莫得」
要得 / 要的         →  统一为「要得」
```

**这是 ASR 的大杀器**：同音异写让模型在多个等价目标间摇摆，既抬高 CER，又浪费建模容量。做法是建一张**方言书写规范映射表**，在生成 jsonl 之前统一 apply。

### 1.5 覆盖度与分布检查

- **字频覆盖**：统计训练集 char-level 频次，方言特有字/词是否每个都出现了足够多次（建议 ≥ 20~50 次）。低频的要么针对性补录，要么上采样。
- **说话人/设备分布**：50h 如果集中在少数说话人、单一录音设备，真实场景（不同麦克风、信道、手机/座机）会打崩。统计一下说话人数、性别、年龄、设备、场景分布。
- **信道不匹配是 CER 高的隐形元凶**，且很容易被忽略。如果训练集全是一个设备录的，测试集是另一个，**先解决这个，收益可能超过所有调参**。

### 1.6 数据增强

官方脚本默认应该已有基础 SpecAugment。额外建议：

| 增强 | 参数 | 说明 |
|---|---|---|
| 语速扰动 | 0.9 / 1.0 / 1.1 | 低成本，方言场景语速方差大 |
| 音量扰动 | ±6 dB | |
| 加噪 | SNR 0~20 dB | **用真实场景噪声**（街道、火锅店、车内、电视背景），不要只用白噪 |
| 混响 | 模拟 RIR | 室内远场场景必备 |
| SpecAugment 调参 | freq_mask / time_mask | 小数据可适当加强 |

增强倍率建议 2~3x，但**验证集不增强**。

---

## 2. 关于语气词 SPECIAL_TOKEN —— 建议重新评估

你正在做的方向有收益，但我建议**顺序调整**，且要注意风险：

### 2.1 建议的优先级

```
文本规范化 > 同音异写收敛 > 语气词统一映射 > 加 SPECIAL_TOKEN
```

前三步做完，CER 往往已经降了一截，且**零风险**。SPECIAL_TOKEN 放在最后作为补充。

### 2.2 加 SPECIAL_TOKEN 的风险

1. **SenseVoice 是非自回归模型**（SANM encoder + 非自回归 decoder）。非自回归模型没有自回归 LM 的"猜词"能力，对稀有 token 更敏感，新加的 token embedding 如果数据不够，学不好反而会拖累。
2. **稀疏性问题**：50h 数据下，如果加了 50~100 个 token，每个平均只出现几百次，新 embedding 训练不充分。
3. **改 tokenizer 的成本**：需要扩 embedding 矩阵，并确认推理侧 tokenizer 同步更新。

### 2.3 更稳妥的替代方案（推荐优先试）

**方案 A：语气词规范化映射（最推荐）**
把语气词统一到固定写法，不加新 token，让模型用已有的字去学：
```
"噻"/"撒"/"啥" → 统一为 "噻"
"哈"/"呵"      → 统一为 "哈"
```

**方案 B：业务无意义则直接删除**
如果语气词对下游任务（如意图理解、质检）没有意义，**直接从训练文本中删掉**，模型就不必输出它们。这是工业 ASR 的常规做法（filler word removal），一举两得：模型压力骤减 + CER 分母里去掉了最难学的部分。

**方案 C：只对极高频语气词加 token**
门槛：词频 ≥ 500 次，且发音稳定、无实义。比如只加 5~10 个最高频的，不要批量加几十上百个。

**方案 D：loss 加权**
把语气词对应的 token 在 loss 中降权，减少模型在这些无语义 token 上的纠结，把容量留给实义词。

> 我的判断：**先做 A 或 B，大概率比加 SPECIAL_TOKEN 收益更高且更快**。A/B 是你的标注规范问题，改脚本重跑一次就行，不用碰 tokenizer 和模型结构。

---

## 3. 训练配置 —— 官方 finetune.sh 默认值对 50h 数据不合适

官方 `finetune.sh` 的默认值（面向大规模数据）：

```bash
++optim_conf.lr=0.0002              # 2e-4，对 50h 微调偏大
++train_conf.max_epoch=50            # 50 epoch，必然过拟合
++train_conf.validate_interval=2000  # 验证太稀疏，选不出好 checkpoint
++train_conf.save_checkpoint_interval=2000
++dataset_conf.batch_size=6000       # token 级 batch，约 60s 音频/batch
++train_conf.keep_nbest_models=20
++train_conf.avg_nbest_model=10      # checkpoint 平均，这个很好
```

### 3.1 建议调整

| 参数 | 官方默认 | 建议 | 理由 |
|---|---|---|---|
| `optim_conf.lr` | 2e-4 | **5e-5 ~ 1e-4** | 50h 数据全参数微调，2e-4 太大，容易破坏预训练表征 |
| `train_conf.max_epoch` | 50 | **10 ~ 20** | 配合早停，50 epoch 纯浪费 |
| `train_conf.validate_interval` | 2000 | **300 ~ 800** | 验证加密才能选到最优 checkpoint（前提：dev 集可靠） |
| `train_conf.save_checkpoint_interval` | 2000 | **300 ~ 800** | 同上 |
| `dataset_conf.batch_size` | 6000 | 保持，或降到 4000 | 小数据可稍微减小 batch 增加更新次数 |
| warmup | - | **显式设置 warmup_steps**（如总步数 5~10%） | 防止训练初期 lr 过大破坏表征 |

### 3.2 冻结 / 分层学习率（LLRD）

50h 数据量，全参数微调容易过拟合。两个低成本方案：

**方案 A：冻结 encoder 底层**
```bash
# 冻结前 N 层 encoder，只训练中上层 + 输出层
++model_conf.freeze_parameters="encoder.encoders.layers.0.*|encoder.encoders.layers.1.*|..."
```
具体层数按实验定：先试冻结前 1/3 层。

**方案 B：分层学习率衰减（LLRD）**
底层 lr 小、顶层 lr 大，衰减系数 0.8~0.9。通常比硬冻结更平滑，收益 1~3 个点。FunASR 的 scheduler 支持自定义，需要改 `train_ds.py` 或传参。

### 3.3 LoRA 微调（强烈建议做一组对比）

50h 数据下，LoRA（r=16~32，target_modules 覆盖 attention 的 q/v）往往比全参数微调**更稳、更不易过拟合、训练更快**。

```python
from peft import LoraConfig, get_peft_model
lora_config = LoraConfig(
    r=32, lora_alpha=32, lora_dropout=0.1,
    target_modules=["q_proj", "v_proj", "out_proj", "linear_q", "linear_v"],
)
```

这是**成本极低、收益明确**的实验点，建议 Full-FT / LoRA 各跑一组对比。

### 3.4 正则化

- Dropout：0.1 → 0.2 / 0.3
- Label smoothing：0.1
- Weight decay：1e-4 ~ 1e-2

### 3.5 Checkpoint 平均 / Model Soup

官方已经开了 `avg_nbest_model=10`，**确保 validate_interval 调密后这个才生效**。另外可以试 weight-space 的 model soup（多个超参配置的最优 ckpt 平均），几乎零成本。

---

## 4. 数据与模型层面的迁移

### 4.1 混合训练（低成本防遗忘）

- **混入一定比例普通话数据**（1:1 或 2:1），防止方言微调导致通用能力崩塌，也相当于正则化。
- **尤其推荐：加入其他西南官话数据**（四川话、贵州话）。西南官话内部互通度极高，公开数据集或自有数据都能用，是**最划算的外部增量**。

### 4.2 重新审视 language token

SenseVoice 的 LID 标签只有 `zh / en / yue / ja / ko / nospeech`，**重庆话没有对应标签**。你现在的 `text_language` 填的是什么？

- 填 `zh`：模型会倾向于把方言发音映射到普通话对应字
- 填 `yue`（粤语）：可能借到模型的"方言建模"能力

**这是一个几乎零成本的对比实验，值得跑一组。** 社区微调方言时两种做法都有。

### 4.3 基座模型对比（重要，但先做 baseline）

在做更多微调优化之前，先确认方向对不对：

```
跑一个 zero-shot Whisper-large-v3 在你的 test set 上，看 CER 是多少？
```

- 如果 zero-shot 就 15% → 说明 SenseVoiceSmall 的容量/预训练对方言不够友好，应考虑换基座或蒸馏
- 如果 zero-shot 40%+ → 说明微调路子是对的，继续深耕数据

同理可测 `paraformer-zh`（FunASR 生态，和 SenseVoice 同源，集成成本低）。

### 4.4 知识蒸馏（放在最后）

如果某大模型（Whisper-large / Qwen2-Audio）在该方言上表现明显更好，可以用它在**无标注方言音频**上生成伪标签，再蒸馏回 SenseVoiceSmall。两步：
1. 先验证老师模型在该方言的 CER 是否显著更低（否则伪标签是噪声）
2. 伪标签做置信度过滤 + 和真实标注混合训练

---

## 5. 解码与后处理（见效快、成本低）

### 5.1 热词 / 上下文偏置

如果业务有固定词表（地名、人名、行业术语、门店名），做成热词列表注入。SenseVoice 推理支持 `hotword` 参数：

```python
res = model.generate(input="audio.wav", hotword="重庆 解放碑 洪崖洞", ...)
```

热词对专有名词召回提升非常明显，且**零训练成本**。

### 5.2 ASR 后纠错（pycorrector）

这正是 pycorrector 的适用场景。分两层：
1. **通用中文纠错**：同音字、语法错误
2. **方言 → 规范文本映射**：自建混淆词表

**注意事项**：
- 纠错会引入新错误，需要**领域词表约束**防止误改
- 建议**只对低置信度片段纠错**，高置信度不动
- 纠错前后都要在 test set 上量 CER，不要让纠错变成负优化

### 5.3 N-best 重打分

SenseVoice 是非自回归模型，原生 N-best 支持有限。如果确实需要，可考虑接外部 LM 对 top-k 重打分，但复杂度较高，放在最后。

### 5.4 VAD 参数调优

长音频的 VAD 切分直接影响 CER（截断会导致 Del 错误激增）：
```python
vad_kwargs={"max_single_segment_time": 30000}  # 按业务调整
```

---

## 参考

- SenseVoice 官方微调：`FunAudioLLM/SenseVoice` 的 `finetune.sh` + `train_ds.py`
- FunASR 降噪模型：`damo/speech_frcrn_ans_cirm_16k`
- FunASR 热词：`model.generate(..., hotword="...")`
- DNSMOS：微软开源无参考语音质量打分（P.808 / P.835）


---

以上回答是AI撰写。

**Comment by alexanderAngle** [comment]
感谢大佬，我之前也没有做过这块，只是前面一段时间根据官方的文档进行训练，每次得到一个字错率数据，对于调优方面一概不知；你的步骤写的非常详细；非常感谢
